src/Ast.ml

```diff
@@ -2393,7 +2395,40 @@ end = struct
     | Ptyp_tuple l -> exposed_right_typ (List.last_exn l)
     | Ptyp_object _ -> true
     | _ -> false
+
+  let parenze_nested_exp {ctx; ast= exp} =
+    let infix_prec ast =
+      match ast with
+      | Exp {pexp_desc= Pexp_apply (e, _); _} when is_infix e ->
+          prec_ast ast
+      | Exp
+          ( { pexp_desc=
+                Pexp_construct
+                  ( {txt= Lident "::"; loc= _}
+                  , Some
+                      { pexp_desc= Pexp_tuple [_; _]
+                      ; pexp_loc= _
+                      ; pexp_attributes= _
+                      ; _ } )
+            ; pexp_loc= _
+            ; pexp_attributes= _
+            ; _ } as exp )
+        when not (is_sugared_list exp) ->
+          prec_ast ast
+      | _ -> None
+    in
+    (* Make the precedence explicit for infix operators *)
+    match (infix_prec ctx, infix_prec (Exp exp)) with
+    | Some (InfixOp0 | ColonEqual), _ | _, Some (InfixOp0 | ColonEqual) ->
+        (* special case for refs update and all InfixOp0 to reduce parens
+           noise *)
+        false
+    | None, _ | _, None -> false
+    | Some p1, Some p2 -> Poly.(p1 <> p2)
 end
 
 include In_ctx
 include Requires_sub_terms
+
+let is_monadic_expr e =
+  match prec_ast (Exp e) with Some InfixOp0 -> true | _ -> false
```

Comment:
This causes a break just after `|>` operators 
```diff
-      |> List.sort
-           ~compare:(Comparable.lift Location.compare_start ~f:Cmt.loc)
+      |>
+      List.sort ~compare:(Comparable.lift Location.compare_start ~f:Cmt.loc)
 ```

Reply:
Yes, that is the new behavior for all these operators, that's why I was filtering a bit (size > 2, checking the last character), so should `|>` be treated like the rest or not?

Reply:
FWIW, I also think this break looks odd. My toy example was using infix operators at the end of each line, a la:
```ocaml
(* Open the repo *)
initialise >>=
(* Perform a subsequent action *)
subsequent_action >|=
```
but when I'm writing this sort of code personally, I keep the infix operators at the start:
```ocaml
x
>>= foo
>>= bar
```
(or the same with `|>`). I don't really mind which of the two OCamlformat chooses to pick (this is probably a reasonably important decision), but I do care about its interaction with comments. i.e., this looks wrong to me:
```ocaml
initialise
>>= (* Perform a subsequent action *)
    subsequent_action 
>|= (* Keep going... *)
    another_action
```
And if OCamlformat chooses to enforce one or the other, there is the question of what to do with input flies that are doing it the other way (comments before/after the infix operator).

Reply:
(this may be an instance where it's worth having an option to select between prepending or appending infix operators)

Reply:
Yes that would be good to have that but it requires a bigger rewrite of the code, the sequences of infix operations are currently sugared into a weird structure that mix different operators (as long as they have the same precedence, e.g. `>>=` and `>|=`), so it's not as simple as it was for sequences of items separated by `;` only.